include/nbl/video/IGPUImage.h

```diff
@@ -60,6 +61,10 @@ class NBL_API IGPUImage : public core::impl::ResolveAlignment<IDeviceMemoryBacke
 		virtual const void* getNativeHandle() const = 0;
 
 	protected:
+		// nullptr if this is not a swapchain image
+		core::smart_refctd_ptr<ISwapchain> m_optionalBackingSwapchain = nullptr;
+		uint32_t m_optionalIndexWithinSwapchain = 0;
+
```

Comment:
wrap this into a `IDeviceMemoryBacked::ICleanup` derivation for the `CVulkanSwapchain`

If you're clever, allocate more memory than needed for the `CVulkanImage`, placement new the `CVulkanImage` and the `ICleanup` over the same memory and override your cleanup's `delete` operator to be a noop (so only one object frees the mem)